exercises/practice/sum-of-multiples/.approaches/from-range/content.md

```diff
@@ -0,0 +1,33 @@
+# Calculating sum by iterating the whole range
+
+```rust
+pub fn sum_of_multiples(limit: u32, factors: &[u32]) -> u32 {
+    (1..limit)
+        .filter(|&n| factors.iter().any(|&factor| factor != 0 && n % factor == 0))
+        .sum()
+}
+```
+
+Instead of implementing the steps in the exercise description, this approach uses another angle:
+
+1. Iterate all numbers between 1 (inclusive) and `limit` (exclusive)
+2. Keep only numbers which have at least one factor in `factors` (automatically avoiding any duplicates)
+3. Calculate the sum of all numbers kept
+
+After creating our range, we use [`filter`][iterator-filter] to keep only matching multiples.
+To detect the multiples, we scan the `factors` and use [`any`][iterator-any] to check if at least
+one is a factor of the number we're checking. ([`any`][iterator-any] is short-circuiting: it stops
+as soon as it finds one compatible factor.)
+
+Finally, once we have the multiples, we can compute the sum easily using [`sum`][iterator-sum].
+
+Although this approach requires scanning the entire range of possible numbers that are multiples,
+it has a few advantages over the technique proposed in the exercise description:
+
+- It is concise and simple to implement and understand
+- It has stable complexity, with a worst case of `O(limit * factors.len())`
```

Comment:
I think you are correct for the _scanning_ part - but then, you have to remove the duplicates and I think this is much harder to quantify. I'm no expert, but since the benchmarks I ran on my machine show a lot of variation depending on how I remove the duplicates, I think that deduplication is actually a big part of the problem.

Reply:
You're right, I didn't think about the deduplication.

With the set-based version, complexity shouldn't change, since deduplication isn't a separate step and inserts are `O(1)` (at least for `HashSet`).

With a vector, we call `sort` and `dedup` afterwards. `dedup` is linear, so that's fine. `sort` is `O(n * log(n))`, where `n` should be... `limit * factors.len()` ? That would actually result in a worse total complexity of `O(limit * factors.len() * log(limit * factors.len()))`.

---

I think it would be valueable to consider this big-picture question about the focus of the article: Whould we discuss performance based on O-notation or the existing concrete benchmarks?

Both can be valid, but right now there's a little bit of a mix which I find confusing. O-notation is given for one approach but not the other. Implementation decisions are made based on concrete benchmark results, though they have worse algorithmic complexity.

I think it would be benefitial to focus only on one aspect and note how it's not the full picture. e.g.

- This is the complexity of approch one, while this is the complexity of approach two. Without benchmarks, approach X may therefore be preferred. However, if you have some knowledge about the shape and size of your input, it can be benefitial to benchmark and optimize according to that.

- Here are some benchmarks comparing the approaches with inputs of shape X and size Y. If you expect your input to look similar to that, you may prefer approach Z. However, if you have no knowledge about your input, the approaches may scale differently and it could be valuable to consider O-notation complexity.

(I realize this is more work and I appreciate the time you already put in, it would be perfectly reasonable to just _not_ do that and get 90% of the value out the door.)